### PR TITLE
perf(linter): reduce mallocs

### PR DESCRIPTION
**Pull request by DonIsaac**
Reduces mallocs via `clones()` (particularly of `Atom`s), and vec allocations.

Affects the following rules:
- [x] eslint/no-dupe-keys
- [x] eslint/no-global-assign
- [x] eslint/no-loss-of-precision _(only partially optimized, needs more work later)_ 
~- [ ] typescript-eslint/ajacent-overload-signatures~ (_will be addressed in following pr_)

### CODE REVIEW

File: crates/oxc_span/src/atom.rs [hunk 1]
```diff
@@ -12,6 +12,10 @@ pub struct Atom(CompactString);
 const BASE54_CHARS: &[u8; 64] = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ$_0123456789";
 
 impl Atom {
+    pub const fn new_inline(s: &'static str) -> Self {
+        Self(CompactString::new_inline(s))
+    }
```
Comment (Boshen) [inline crates/oxc_span/src/atom.rs:17]: This is not used yet.

File: crates/oxc_linter/src/rules/eslint/no_dupe_keys.rs [hunk 2]
```diff
@@ -42,29 +43,50 @@ declare_oxc_lint!(
...
+    match key {
+        PropertyKey::Identifier(ident) => Some(calculate_hash(&ident)),
+        PropertyKey::PrivateIdentifier(_) => None,
+        PropertyKey::Expression(expr) => match expr {
+            Expression::StringLiteral(lit) => Some(calculate_hash(&lit.value)),
+            // note: hashes won't work as expected if these aren't strings. Save
+            // NumberLiteral I don't think this should be too much of a problem
+            // b/c most people don't use `null`, regexes, etc. as object
+            // property keys when writing real code.
+            Expression::RegExpLiteral(lit) => Some(calculate_hash(&lit.regex.to_string())),
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/eslint/no_dupe_keys.rs:78]: Yeah we need to fix  the `to_string` at a later time.

Comment (DonIsaac) [reply crates/oxc_linter/src/rules/eslint/no_dupe_keys.rs:78]: +1

File: crates/oxc_span/src/atom.rs (hunk 1 above)
Comment (DonIsaac) [reply crates/oxc_span/src/atom.rs:17]: Sorry, this got committed from my working branch

**Pull request merged**